File: backend/app/schemas/analytics.py
from pydantic import BaseModel


class AnalyticsSummary(BaseModel):
    total_applications: int
    active_applications: int
    response_rate: float  # percentage
    offer_count: int
    rejection_count: int
    avg_days_to_response: float | None


class StageFunnelEntry(BaseModel):
    stage: str
    count: int
    percentage: float


class StageFunnelResponse(BaseModel):
    entries: list[StageFunnelEntry]


class TimelineBucket(BaseModel):
    period: str  # "2026-03-17" or "2026-W12"
    count: int


class TimelineResponse(BaseModel):
    buckets: list[TimelineBucket]
    granularity: str  # "daily" or "weekly"


class ResponseRateBucket(BaseModel):
    period: str
    total: int
    responded: int
    rate: float


class ResponseRateResponse(BaseModel):
    buckets: list[ResponseRateBucket]


class StageDistributionEntry(BaseModel):
    stage: str
    count: int
    percentage: float


class StageDistributionResponse(BaseModel):
    entries: list[StageDistributionEntry]

File: frontend/src/api/analytics.ts
import api from "@/api/client"

export interface AnalyticsSummary {
  total_applications: number
  active_applications: number
  response_rate: number
  offer_count: number
  rejection_count: number
  avg_days_to_response: number | null
}

export interface StageFunnelEntry {
  stage: string
  count: number
  percentage: number
}

export interface TimelineBucket {
  period: string
  count: number
}

export interface ResponseRateBucket {
  period: string
  total: number
  responded: number
  rate: number
}

export interface StageDistributionEntry {
  stage: string
  count: number
  percentage: number
}

export async function getAnalyticsSummary(): Promise<AnalyticsSummary> {
  const { data } = await api.get<AnalyticsSummary>("/analytics/summary")
  return data
}

export async function getStageFunnel(): Promise<StageFunnelEntry[]> {
  const { data } = await api.get<{ entries: StageFunnelEntry[] }>("/analytics/funnel")
  return data.entries
}

export async function getApplicationTimeline(): Promise<{
  buckets: TimelineBucket[]
  granularity: string
}> {
  const { data } = await api.get("/analytics/timeline")
  return data
}

export async function getResponseRates(): Promise<ResponseRateBucket[]> {
  const { data } = await api.get<{ buckets: ResponseRateBucket[] }>("/analytics/response-rates")
  return data.buckets
}

export async function getStageDistribution(): Promise<StageDistributionEntry[]> {
  const { data } = await api.get<{ entries: StageDistributionEntry[] }>(
    "/analytics/stage-distribution"
  )
  return data.entries
}
